### PR TITLE
Use pallet names to name enum variants

### LINKED ISSUE #8862: Change construct_runtime! macro to use pallet names instead of pallet paths for enum variant identifiers
Currently, during macro expansion of `construct_runtime!`, the enums that we generate uses the pallet path rather than the pallet name as a component for the enum variants:

```rust
construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		Foo: bar::{Pallet, Call, Storage, Event, Origin},
	}
);

// This would generate the following Event enum for example:

pub enum Event {
	bar(bar::Event)
}
```

Ideally, we would use the pallet name as specified in the `construct_runtime!` macro so that we generate the following instead:

```rust
pub enum Event {
	Foo(bar::Event)
}
```

In other words, audit calls to `PalletPath::mod_name`, and see if they can be replaced by the pallet name. `mod_name` should ideally not exist.

Files: `frame/support/procedural/src/construct_runtime/mod.rs `, `frame/support/procedural/src/construct_runtime/expand/event.rs`, `frame/support/procedural/src/construct_runtime/expand/event.rs`, `frame/support/procedural/src/construct_runtime/expand/origin.rs`

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #8862.

polkadot companion: paritytech/polkadot#3194

The variants of `OriginCaller` and `Event` enums and the field names of `GenesisConfig` are going to be renamed under a new scheme.

Firstly, `OriginCaller` and `Event` enum variants are going to be named by just their pallet name as specified in `construct_runtime`. There is no more need to use the crate/module name of the pallet and concatenating it with the instance number.

`GenesisConfig` fields will also be renamed according to the pallet name, but with a twist -- pallet names are PascalCase by convention, and the new naming scheme converts all of them into snake_case when generating the field names.

The conversion from the PascalCasePalletName to snake_case_pallet_name is done by replacing all uppercase letters in the PascalCasePalletName with an underscore, followed by the lowercase equivalent, except in cases where the uppercase letter is at the beginning of the word, where only lowercasing is sufficient and no underscores are necessary, e.g. if I have `MyPallet` as the pallet name, the field name for it in `GenesisConfig` is `my_pallet`. Most notably, if I have `BridgePoA`, then the equivalent field name is `bridge_po_a`. For this reason, it's recommended (not just by me but is also Rust convention) to treat all abbreviations as proper words, i.e. use Poa instead of PoA, use Pos instead of PoS and so on.

### CODE REVIEW

File: frame/support/procedural/src/construct_runtime/expand/config.rs [hunk 1]
```diff
@@ -32,11 +33,11 @@ pub fn expand_outer_config(
 	for decl in pallet_decls {
 		if let Some(pallet_entry) = decl.find_part("Config") {
 			let config = format_ident!("{}Config", decl.name);
-			let mod_name = decl.pallet.mod_name();
+			let pallet_name = &decl.name.to_string().to_snake_case();
 			let field_name = if let Some(inst) = decl.instance.as_ref() {
-				format_ident!("{}_{}", mod_name, inst)
+				format_ident!("{}_{}", pallet_name, inst.to_string().to_snake_case())
```
Comment (KiChjang) [inline frame/support/procedural/src/construct_runtime/expand/config.rs:38]: This here is technically a breaking change as well, unsure if it's worth?

**Review by coriolinus — APPROVED** [review]
Change looks good in and of itself. I'm not sure about the potential side-effects which we'll experience after this is merged. Will this have any effects visible to an end-user?

File: frame/support/procedural/src/construct_runtime/expand/config.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/procedural/src/construct_runtime/expand/config.rs:38]: we don't need to concatenate with the instance name, the pallet name is already unique, and I think it worth it, the name is better than the path to identify the pallet

File: frame/support/procedural/src/construct_runtime/expand/event.rs [hunk 2]
```diff
@@ -77,43 +75,45 @@ pub fn expand_outer_event(
...
-	path: &PalletPath,
+	pallet: &Pallet,
 	instance: Option<&Ident>,
 	pallet_event: &TokenStream,
 ) -> TokenStream {
-	let mod_name = path.mod_name();
+	let pallet_name = &pallet.name;
 	let variant = if let Some(inst) = instance {
-		format_ident!("{}_{}", mod_name, inst)
+		format_ident!("{}_{}", pallet_name, inst)
```
Comment (thiolliere) [inline frame/support/procedural/src/construct_runtime/expand/event.rs:114]: same no need for instance to be concatenated here

**Comment by thiolliere** [comment]
we should be careful that metadata doesn't change in a way that can break UI, I haven't check it yet

**Comment by KiChjang** [comment]
> we should be careful that metadata doesn't change in a way that can break UI, I haven't check it yet

It seems to me that it doesn't touch anything related to metadata. Firstly, the logic in `expand/metadata.rs` is untouched, and secondly, the metadata code doesn't rely upon the naming scheme for the enum variants of `OriginCaller`, `Event` or in the case of `GenesisConfig`, the fields. I've also ran the `test_metadata` test in `construct_runtime.rs` to make sure I didn't run into any weird issues, and it passes successfully.

**Review by thiolliere — APPROVED** [review]
PR is good, I think the change makes complete sense and is future-proof, as you said https://github.com/paritytech/substrate/issues/8862 the change shouldn't be difficult to fix, just renaming some variants and fields.

Maybe if it is consider a too much annoying change, we can make it optional, is there people worrying about it ?

**Comment by emostov** [comment]
> It seems to me that it doesn't touch anything related to metadata. 

I can't tell myself just by reading the changes - but would it be worth diffing the metadata on master vs this branch. (I think the metadata diffing is outlined here would be sufficient: https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines)

**Comment by shawntabrizi** [comment]
I agree with @emostov just to make sure

**Review by shawntabrizi — APPROVED** [review]
needs companion it seems

**Comment by KiChjang** [comment]
So I checked for differences between the metadata generated by this PR and master via `diff`, and I found no difference. I think it's safe to say no metadata was harmed during the course of this software surgery.

**Comment by thiolliere** [comment]
I think it worth having a precise explanation of the variant/fields renaming in the top message of this PR, so that people can upgrade more easily

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
Does this change the serde implementation of the genesis config ? could this be an issue ?

**Comment by bkchr** [comment]
> This change the serde implementation of the genesis config, could this be an issue ?

Only for non raw genesis configs.

**Comment by xlc** [comment]
Just to confirm, this is not a runtime breaking change right? Just renames.

**Comment by KiChjang** [comment]
Correct, nothing about the runtime is changed, except for the `GenesisConfig` fields that require renaming.

**Pull request merged**